File: v1/commands/TeamsCommand.rb
class TeamsCommand < BaseCommand
  desc "list", "Show a list of NHL teams."
  def list(teamId=nil)
    initTask()

    if teamId.nil?
      showTeams()
    else
      showRosterForTeamId(teamId)
    end
  end

  default_task :list

  no_tasks do
    def showTeams
      Logger.taskTitle "Roster - Show all teams"

      teams = @dbService.getAllTeams()

      proportionPadding = 10
      Logger.header([
        LoggerColumn.id(),
        LoggerColumn.name(),
        LoggerColumn.custom("P %", proportionPadding)
      ])

      teams.each do |team|
        teamRoster = @dbService.getRosterForTeam(team)
        nbPlayersRoster = teamRoster.count()
        nbPlayerInDb = 0

        teamRoster.each do |playerId|
          player = @dbService.getPlayerForId(playerId)

          unless player.nil?
            nbPlayerInDb += 1
          end
        end

        proportion = "#{nbPlayerInDb} / #{nbPlayersRoster}"
        Logger.row([
          LoggerColumn.id(team.id),
          LoggerColumn.name(team.name),
          LoggerColumn.custom(proportion, proportionPadding)
        ])
      end

      Logger.taskEnd
    end

    def showRosterForTeamId(id)
      team = @dbService.getTeamForId(id)
      roster = @dbService.getRosterForTeam(team)

      Logger.taskTitle "Roster - Team #{id} #{team.name}"

      Logger.header([
        LoggerColumn.id(),
        LoggerColumn.name()
      ])

      roster.each do |playerId|
        player = @dbService.getPlayerForId(playerId)
        fullName = player.nil? ? "" : player.fullName
        Logger.row([
          LoggerColumn.id(playerId),
          LoggerColumn.name(fullName)
        ])
      end

      Logger.taskEnd
    end
  end
end

File: src/main/kotlin/ca/sebleclerc/hockeydata/commands/TeamsCommand.kt
package ca.sebleclerc.hockeydata.commands

import ca.sebleclerc.hockeydata.DI
import ca.sebleclerc.hockeydata.helpers.Logger
import ca.sebleclerc.hockeydata.helpers.LoggerColumn
import ca.sebleclerc.hockeydata.models.CacheStep
import com.github.ajalt.clikt.core.CliktCommand
import com.github.ajalt.clikt.parameters.arguments.argument

class TeamsCommand(private val di: DI) : CliktCommand(name = "teams") {
  override fun run() {
    Logger.taskTitle("Teams")

    val steps: MutableList<CacheStep> = di.database.getAllTeams().map { CacheStep.CacheTeamRoster(it) }.toMutableList()
    steps.add(0, CacheStep.Teams())

    Logger.enabled = false

    di.cache.cache(
      steps,
      force = false
    )

    Logger.enabled = true

    val proportionPadding = 10
    Logger.header(
      LoggerColumn.ID(),
      LoggerColumn.Name(),
      LoggerColumn.Custom("P %", proportionPadding)
    )

    di.database.getAllTeams().forEach { team ->
      val roster = di.database.getRosterForTeam(team.id)
      val nbPlayersInRoster = roster.count()
      var nbPlayerInDB = 0

      roster.forEach { playerId ->
        val player = di.database.getPlayerForId(playerId)
        if (player != null) nbPlayerInDB += 1
      }

      val proportion = "$nbPlayerInDB / $nbPlayersInRoster"

      Logger.row(
        LoggerColumn.ID(team.id),
        LoggerColumn.Name(team.name),
        LoggerColumn.Custom(proportion, proportionPadding)
      )
    }

    Logger.taskEnd()
  }
}
